src/operator/tensor/elemwise_binary_broadcast_op.h

```diff
@@ -134,24 +137,122 @@ inline int BinaryBroadcastShapeCompact(const TShape& lshape, const TShape& rshap
 }
 
 namespace mxnet_op {
-template<int ndim, typename DType, typename OP>
-struct binary_broadcast_kernel {
+
+/*!
+ * \brief Type-level specialization base class for binary_broadcast_kernel
+ * \tparam DType
+ */
+template <typename DType> struct tunable_binary_broadcast_kernel;
+
+template<int ndim, typename DType, typename OP, bool is_tuning = false>
+struct binary_broadcast_kernel : public tunable_binary_broadcast_kernel<DType> {
+  /*! \brief Map function for binary_broadcast_kernel */
   MSHADOW_XINLINE static void Map(int base, int length, OpReqType req,
                                   const Shape<ndim>& lstride, const Shape<ndim>& rstride,
                                   const Shape<ndim>& oshape, DType* lhs, DType* rhs,
-                                  DType* out, int lsize, int rsize) {
-    Shape<ndim> coord = unravel(base, oshape);
-    index_t lidx = dot(coord, lstride);
-    index_t ridx = dot(coord, rstride);
-    KERNEL_ASSIGN(out[base], req, OP::Map(lhs[lidx], rhs[ridx]));
-    // starts from 1 to avoid extra inc at end of loop
-    for (int i = 1; i < length; ++i) {
-      inc(&coord, oshape, &lidx, lstride, &ridx, rstride);
-      KERNEL_ASSIGN(out[base+i], req, OP::Map(lhs[lidx], rhs[ridx]));
+                                  DType* out) {
+    if (req != kNullOp) {
+      Shape <ndim> coord = unravel(base, oshape);
+      auto lidx = static_cast<index_t>(dot(coord, lstride));
+      auto ridx = static_cast<index_t>(dot(coord, rstride));
+      if (!is_tuning) {
+        KERNEL_ASSIGN(out[base], req, OP::Map(lhs[lidx], rhs[ridx]));
+      }
+      // starts from 1 to avoid extra inc at end of loop
+      for (int i = 1; i < length; ++i) {
+        inc(&coord, oshape, &lidx, lstride, &ridx, rstride);
+        // When tuning, don't actually run the op, since it's not going to be tuned against
+        // the actual op we'll eventually be using
+        if (!is_tuning) {
+          KERNEL_ASSIGN(out[base + i], req, OP::Map(lhs[lidx], rhs[ridx]));
+        }
+      }
     }
   }
+
+  /*!
+   * \brief Decide whether to use OpenMP parallelization
+   * \tparam Args Variable number and types of arguments (passed same args as Map())
+   * \param N Number of iterations
+   * \param thread_count Number of OMP threads available
+   * \param req Req type (i.e. kNullOp, kWriteTo, kAddTo, etc)
+   * \param args remaining (unused) arguments
+   * \return true if OMP parallelization should be used for the N iterations
+   */
+  template<typename ...Args>
+  static bool UseOMP(const size_t N, const size_t thread_count, OpReqType req, Args... args) {
```

Comment:
This looks wildly too complicated for a broadcasting kernel

Reply:
“Wildly”? Ok so won’t support broadcast. I’ll remove.

Reply:
broadcast support removed